GH-414 - Documented explanation (#20)
chb0github/pyrsia: docs/blockchain/why-us-aleph.md

- Feature Name: (fill me in with a unique ident, `my_awesome_feature`)
- Start Date: (fill me in with today's date, YYYY-MM-DD)
- RFC PR: [pyrsia/pyrsia#0000](https://github.com/pyrsia/pyrsia/pull/0000)
- Pyrsia Issue: [pyrsia/pyrsia#0000](https://github.com/pyrsia/pyrsia/issues/0000)

# Summary

> Why we chose to use Aleph for our consensus engine

Reviewer: I read this like there's more meat here

```suggestion
Exploration, reasoning and choice for a consensus engine to build on top of
```

# Motivation

One of Pyrsia's main design decisions is to use a blockchain for its underlying datastore. The data structure(s) of a
blockchain are plenty challenging but nothing mind bending. However, the real challenge to blockchain is consensus;
how do you get a bunch of strangers to trust each other and agree, when some of them might be corrupt and you could
have any number of faults in the system?

> ## Core Requirements

Reviewer: I think we should link the google doc

Author: Oh? Yours? Let me see if I can find it

- No currency. We wish to avoid legal problems
- Short TTF
- Apache 2 or compatible license (To avoid legal problems)
- Active community development (necessary for good support)

Given the complexity of consensus and the data structures used to manage a blockchain, it's advisable to find existing
code that can be leveraged; stand on the shoulders of giants, as the expression goes. However, this led us into a
contradiction with another requirement for Pyrsia; a legal and regulatory one: No currency. Each of these existing
libraries that implement blockchain consensus and data structures all assume there is a currency involved. After all,
this is the biggest reason people have gravitated toward blockchain: Bitcoin and Ethereum.

# Prior art

## Aleph

> There was only 1 library out there that exclusively did consensus and nothing else: [Aleph](https://github.com/aleph-zero-foundation/AlephBFT).

Reviewer: ```suggestion
There was one library  in particular that stood out for focusing on consensus and nothing else: [Aleph](https://github.com/aleph-zero-foundation/AlephBFT).
```

Author: That's kind of fussy, don't you think?

Also, the implementation was simple as only basic traits needed to be implemented
1. Sign/verify

> 2. I/O

Reviewer: I mean theres three listed https://github.com/aleph-zero-foundation/AlephBFT#using-the-crate, might as well just copy the names and link to rust docs

- Feature Name: (fill me in with a unique ident, `my_awesome_feature`)
- Start Date: (fill me in with today's date, YYYY-MM-DD)
- RFC PR: [pyrsia/pyrsia#0000](https://github.com/pyrsia/pyrsia/pull/0000)
- Pyrsia Issue: [pyrsia/pyrsia#0000](https://github.com/pyrsia/pyrsia/issues/0000)

# Summary

Why we chose to use Aleph for our consensus engine

# Motivation

One of Pyrsia's main design decisions is to use a blockchain for its underlying datastore. The data structure(s) of a
blockchain are plenty challenging but nothing mind bending. However, the real challenge to blockchain is consensus;
how do you get a bunch of strangers to trust each other and agree, when some of them might be corrupt and you could
have any number of faults in the system?

> # Our core requirements #

Reviewer: ```suggestion
## Core Requirements
```

> - No currency. We wish to avoid legal problems

Reviewer 2: Cryptocurrency and consensus algorithm are different concepts.

Author: True, and I don't believe that I conflated them. Although, One of the mechanisms to encourage concensus participation is "Proof of Work" - and, participants are paid for winning the contest in coinage. So, they are, by design, linked together.

Reviewer: that's a little bit of a mischaracterization... having a reward mechanism is not at all related to PoW.

"incentivising" can also be a negative affect like "Proof of Burn"

I think it's better to phrase this

```suggestion
- Avoid any notion of monetary assets (e.g balance, account, tokens)
```

> - High TPS and TTF

Reviewer: TPS does not matter and TTF should be small

```suggestion
- Short TTF
```

Reviewer 2: Please provide the full form of an abbreviation at first use

> - Apache 2 or compatible license

Reviewer: There was a few other items like active community development as I recall

Author: You're right - let me update

Given the complexity of consensus and the data structures used to manage a blockchain, it's advisable to find existing
code that can be leveraged; stand on the shoulders of giants, as the expression goes. However, this led us into a
contradiction with another requirement for Pyrsia; a legal and regulatory one: No currency. Each of these existing
libraries that implement blockchain consensus and data structures all assume there is a currency involved. After all,
this is the biggest reason people have gravitated toward blockchain: Bitcoin and Ethereum.

# Prior art

> ## Aleph ##

Reviewer: ```suggestion
## Aleph
```

There was only 1 library out there that exclusively did consensus and nothing else: [Aleph](https://github.com/aleph-zero-foundation/AlephBFT).
Also, the implementation was simple as only basic traits needed to be implemented
1. Sign/verify
2. I/O

In addition, because `verify` and `sign` as so cleanly isolated, we are able to use proof of authority. Which is simple and gives enterprise IT admins warm fuzzy feelings.

> Finally, it's block uses very well tread data structures, that have been proven repeatedly.

Reviewer: There's no structure definitions 🤔

Author: Do I need to discuss and describe Merkel trees?

> ## Substrate ##

Reviewer: ```suggestion
## Substrate
```

Another such tech [substrate](https://substrate.io/) did much more, and had a compatible license, but had the
assumption of a currency. We had considered setting transaction costs and gas to 0 (thus neutralising block currency), but even this was deemed too much.

> ## Tendermint ##

Reviewer: ```suggestion
## Tendermint
```

The last tech truly considered was [tendermint](https://tendermint.com/). However, their examples wouldn't compile and
when reaching out to community, after taking extreme measures to get it working, the response was ["We don't have the
bandwidth to maintain docs in as many languages set of docs so we're planning on removing them"](https://github.com/tendermint/tendermint/issues/7743#issuecomment-1028025629).
In addition, a core part of their platform is _not_ license compatible.

# Unresolved questions

There are 2 outstanding questions:

> 1. What sort of key or certificate should we use to sign things? X509, which is hierarchical or simple public/private key which are quicker and easier to deploy

Reviewer: I know you love X509 but they aren't aligned with blockchain's zero-trust, I'd rather not have mentions of it since people existing knowledge might cloud there ideas

```suggestion
```

Author: But, it is an unresolved question. Pyrsia is "zero" trust in the sense that we make "zero" trust assertions. I actually spoke with a friend of mine who is a long time sys admin about this choice. He is on the fence - to me this isn't so clear cut. I think bringing it up to the group at large is important so it can't be claimed we didn't discuss and think about it

> 2. Should metadata be cooked right into the block and transactions themselves or should metadata be an opaque part of the payload?

Reviewer: There's no block metadata.... it's too confusing with the "package metadata" which we have no clue about packages let alone what else we might invent

```suggestion
```

Author: There is no called out structure for block meta-data, but there is an understanding that there will be meta-data surrounding each artifact/transaction. It was discussed quite a lot in Atlanta and pyrsia was directly compared to another OSS project I am trying to find. 

You and I have had this discussion: Will meta data exist explicitly in our block/transaction or will users just jam data in the block payload and be left to their own devices to determine what that means and if it's meta or not?

This is still an open discussion and I am sure the team will want to have some say since they will need it. I will see about scheduling something to discuss with the team.

# Motivation

One of Pyrsia's main design decisions is to use a blockchain for its underlying datastore. The data structure(s) of a
blockchain are plenty challenging but nothing mind bending. However, the real challenge to blockchain is consensus;
how do you get a bunch of strangers to trust each other and agree, when some of them might be corrupt and you could
have any number of faults in the system?

## Core Requirements
- Avoid any notion of monetary assets (e.g balance, account, tokens)
- Short TTF
- Apache 2 or compatible license (To avoid legal problems)
- Active community development (necessary for good support)

Given the complexity of consensus and the data structures used to manage a blockchain, it's advisable to find existing
code that can be leveraged; stand on the shoulders of giants, as the expression goes. However, this led us into a
contradiction with another requirement for Pyrsia; a legal and regulatory one: No currency. Each of these existing
libraries that implement blockchain consensus and data structures all assume there is a currency involved. After all,
this is the biggest reason people have gravitated toward blockchain: Bitcoin and Ethereum.

> # Prior art

Reviewer: I feel like we are missing the "process"

> we started with ... to learn ... we looked at ... to find out ... then settled on ... because it...

## Aleph

There was one library in particular that stood out for focusing on consensus and nothing else: [Aleph](https://github.com/aleph-zero-foundation/AlephBFT).
Also, the implementation was simple as only 3 [basic traits needed to be implemented](https://github.com/aleph-zero-foundation/AlephBFT#using-the-crate)
1. Sign/verify
2. I/O (distinct from network IO)
3. Network

A more detailed write-up is available [here](https://docs.google.com/presentation/d/1hJIa9Ht-s99CuHGou6Ra9Y7Lkk2OIV9OjxM4wVDFcao/edit#slide=id.g1156090f3c7_1_125)

> In addition, because `verify` and `sign` as so cleanly isolated, we are able to use proof of authority. Which is simple and gives enterprise IT admins warm fuzzy feelings.

Reviewer: I personally dislike the commentary

Author: I can change it to be a bit more neutral
